pkg/job/job.go

```diff
@@ -0,0 +1,252 @@
+package job
+
+import (
+	"os"
+	"os/exec"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/runtime/protoiface"
+
+	"github.com/MinhNghiaD/jobworker/api/worker/proto"
+	"github.com/MinhNghiaD/jobworker/pkg/log"
+)
+
+// Job is managed by the worker for the execution of linux process
+type Job interface {
+	// Start the job in the background
+	Start() error
+	// Stop the job
+	Stop(force bool) error
+	// Query the current status of the job
+	Status() *proto.JobStatus
+}
+
+// TODO Handle QUEUING state when job queue and scheduler is implemented
+// For now, the job will be start immediately after created
+// Process waiting to be executed
+
+// Impl is the implementation of the Job interface
+type Impl struct {
+	id         uuid.UUID
+	cmd        *exec.Cmd
+	logger     *log.Logger
+	owner      string
+	state      proto.ProcessState
+	stateMutex sync.RWMutex
+	exitChan   chan bool
+}
+
+// newJob creates a new job, associated with a unique ID and a Command, and the owner who requests its
+func newJob(ID uuid.UUID, cmd *exec.Cmd, logger *log.Logger, owner string) (Job, error) {
+	return &Impl{
+		id:       ID,
+		cmd:      cmd,
+		logger:   logger,
+		owner:    owner,
+		state:    proto.ProcessState_RUNNING,
+		exitChan: make(chan bool, 1),
+	}, nil
+}
+
+// Start runs job in the background
+func (j *Impl) Start() error {
+	// IO Mapping
+	stdoutLog := j.logger.Entry.WithField("source", "stdout").Writer()
+	stderrLog := j.logger.Entry.WithField("source", "stderr").Writer()
```

Comment:
Can you explain why you are using `logrus` here? I don't recall seeing `logrus` in the design document and don't quite understand the need to go through `logrus` here? Why can't you write straight to files on disk?

Reply:
Hi, It was not as part of the documentation but during the discussion with Andrew during the review of the documentation, I think it would be cool to identify with log entry comes from stdout and which comes from stderr. Andrew suggested using 2 different files to simplify the output but then merging them back in the right order before returning the log to the user is quite problematic. With logrus we can format the log entry in JSON and add a field to specify the source as I did here. Plus logrus logger comes with a built-in mutex and Hook which can be used for receiving writing events that are quite useful for streaming. About the utilization of logrus Hook for streaming, I explained in the messages that I sent yesterday if you are interested.

Reply:
Also - what happens if the output is not lines of text?

Reply:
As I understand logrus logger implementation, for each writes to the logger, the entire data is encoded into the "msg" field of a  JSON object, this object then is written to the file as a line. So as long as a series of output is fired in a write, it is considered as an entry by logrus. And with Hook, at each write, I captured the offset and the size of that entry in the file. So I don't think that the format of output of the process will affect the program at all

Reply:
Okay, I see what you are doing, but not sure I agree with the approach. Using logrus feels like the wrong approach here, nothing wrong with going with the simpler two file approach. Ordering is not a requirement here, you can just return two different streams, that's fine.

Reply:
Thanks

Reply:
I don't think that logrus added more complexity to the project. Since it just like using logger as a file abstraction without doing anything more. It is true that logrus make the file a little larger by adding more info, but it also creates a well-structured file. I think that it is a sensible tradeoff